File: apps/backend/src/common/repository/base.repository.ts
import { DeepPartial, FindOptionsWhere, ObjectLiteral, Repository } from 'typeorm';

/**
 * BaseRepository<T> — generic repository pattern.
 *
 * Phase 2 scope: a reusable abstraction over TypeORM's Repository<T>
 * that feature repositories can extend once real entities exist. It
 * intentionally has no knowledge of any specific domain model (loan,
 * user, document, etc.) — those concrete repositories/entities are
 * out of scope until database schema work begins.
 *
 * Example future usage:
 *   @Injectable()
 *   export class LoanRepository extends BaseRepository<LoanEntity> {
 *     constructor(@InjectRepository(LoanEntity) repo: Repository<LoanEntity>) {
 *       super(repo);
 *     }
 *   }
 */
export abstract class BaseRepository<T extends ObjectLiteral> {
  protected constructor(protected readonly repository: Repository<T>) {}

  async findAll(): Promise<T[]> {
    return this.repository.find();
  }

  async findOneById(id: string | number): Promise<T | null> {
    return this.repository.findOneBy({ id } as unknown as FindOptionsWhere<T>);
  }

  async findOne(where: FindOptionsWhere<T>): Promise<T | null> {
    return this.repository.findOneBy(where);
  }

  async create(data: DeepPartial<T>): Promise<T> {
    const entity = this.repository.create(data);
    return this.repository.save(entity);
  }

  async update(id: string | number, data: DeepPartial<T>): Promise<T | null> {
    await this.repository.update(id, data as never);
    return this.findOneById(id);
  }

  async delete(id: string | number): Promise<boolean> {
    const result = await this.repository.delete(id);
    return (result.affected ?? 0) > 0;
  }
}

File: packages/shared-flutter/lib/src/repository/base_repository.dart
import '../network/api_client.dart';
import '../network/api_result.dart';

/// Repository pattern — base class for feature repositories.
///
/// Phase 2 scope: a generic abstraction over [ApiClient] that future
/// feature repositories (e.g. a `LoanRepository` in a later phase)
/// will extend. It intentionally knows nothing about any specific
/// domain model — no business logic or API endpoints are defined here.
abstract class BaseRepository {
  BaseRepository(this.apiClient);

  final ApiClient apiClient;

  /// Convenience wrapper so subclasses can write:
  ///   Future<ApiResult<Loan>> getLoan(String id) =>
  ///     get('/loans/$id', mapper: Loan.fromJson);
  Future<ApiResult<T>> get<T>(
    String path, {
    required T Function(dynamic data) mapper,
    Map<String, dynamic>? queryParameters,
  }) {
    return apiClient.request<T>(
      (dio) => dio.get(path, queryParameters: queryParameters),
      mapper: mapper,
    );
  }

  Future<ApiResult<T>> post<T>(
    String path, {
    required T Function(dynamic data) mapper,
    Object? body,
  }) {
    return apiClient.request<T>(
      (dio) => dio.post(path, data: body),
      mapper: mapper,
    );
  }

  Future<ApiResult<T>> patch<T>(
    String path, {
    required T Function(dynamic data) mapper,
    Object? body,
  }) {
    return apiClient.request<T>(
      (dio) => dio.patch(path, data: body),
      mapper: mapper,
    );
  }
}
